packages/client/src/features/tools/resize-key-tool.ts

```diff
@@ -0,0 +1,189 @@
+            const resizeElement = { id: element.id, bounds: { x, y, width, height } } as SModelElement & SParentElement & Resizable;
+            return ChangeBoundsOperation.create([toElementAndBounds(resizeElement)]);
+        }
+
+        return undefined;
+    }
+
+    protected reduceSize(element: SelectableBoundsAware): Action | undefined {
+        const { x, y, width: oldWidth, height: oldHeight } = element.bounds;
+        const width = oldWidth - this.grid.x;
+        const height = oldHeight - this.grid.y;
+
+        if (this.isValidBoundChange(element, { x, y }, { width, height })) {
+            const resizeElement = { id: element.id, bounds: { x, y, width, height } } as SModelElement & SParentElement & Resizable;
+            return ChangeBoundsOperation.create([toElementAndBounds(resizeElement)]);
+        }
+
+        return undefined;
+    }
+
+    protected minSize(element: SelectableBoundsAware): Action | undefined {
+        const { x, y } = element.bounds;
+        const width = minWidth(element);
+        const height = minHeight(element);
+
+        if (this.isValidBoundChange(element, { x, y }, { width, height })) {
+            const resizeElement = { id: element.id, bounds: { x, y, width, height } } as SModelElement & SParentElement & Resizable;
+            return ChangeBoundsOperation.create([toElementAndBounds(resizeElement)]);
+        }
+
+        return undefined;
+    }
+
+    protected isValidBoundChange(element: SelectableBoundsAware, newPosition: Point, newSize: Dimension): boolean {
+        return this.isValidSize(element, newSize) && this.isValidMove(element, newPosition);
+    }
+
+    protected isValidSize(element: SelectableBoundsAware, size: Dimension): boolean {
+        return isValidSize(element, size);
+    }
+
+    protected isValidMove(element: SelectableBoundsAware, newPosition: Point): boolean {
+        return isValidMove(element, newPosition, this.tool.movementRestrictor);
+    }
+
+    protected getSelectedElements(): SelectableBoundsAware[] {
+        const index = this.tool.selectionService.getModelRoot().index;
+        const selectedElements = this.tool.selectionService.getSelectedElementIDs();
+        return getElements(index, selectedElements, isSelectableAndBoundsAware);
+    }
+
+    protected matchesIncreaseSizeKeystroke(event: KeyboardEvent): boolean {
+        return event.key === '+';
+    }
+
+    protected matchesActivateResizeModeKeystroke(event: KeyboardEvent): boolean {
+        return matchesKeystroke(event, 'KeyR', 'alt');
+    }
+
+    protected matchesDeactivateResizeModeKeystroke(event: KeyboardEvent): boolean {
+        return matchesKeystroke(event, 'Escape');
+    }
+
+    protected matchesMinSizeKeystroke(event: KeyboardEvent): boolean {
+        return matchesKeystroke(event, 'Digit0', 'ctrl');
+    }
+
+    protected matchesReduceSizeKeystroke(event: KeyboardEvent): boolean {
+        return matchesKeystroke(event, 'Minus');
+    }
```

Comment:
As mentioned in the move/zoom PR, it would be great to support numpad keys as well: https://github.com/eclipse-glsp/glsp-client/pull/241#discussion_r1189568526